network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,356 @@
+        // parse the server's response
+        // TODO: security logging here? (mimoo)
+        let (_, session) = self
+            .0
+            .finalize_connection(initiator_state, &server_response)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // finalize the connection
+        Ok(NoiseStream::new(socket, session))
+    }
+
+    pub async fn accept<TSocket>(
+        &self,
+        mut socket: TSocket,
+        anti_replay_timestamps: Option<Arc<RwLock<AntiReplayTimestamps>>>,
+        trusted_peers: Option<&Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>>,
+    ) -> io::Result<NoiseStream<TSocket>>
+    where
+        TSocket: AsyncRead + AsyncWrite + Unpin,
+    {
+        // receive the initiation message
+        let mut client_init_message = [0u8; noise::handshake_init_msg_len(PAYLOAD_SIZE)];
+        socket.read_exact(&mut client_init_message).await?;
+
+        // parse it
+        let (their_public_key, handshake_state, payload) = self
+            .0
+            .parse_client_init_message(&[], &client_init_message)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // make sure the public key is a validator before continuing (if we're in the validator network)
+        if let Some(trusted_peers) = trusted_peers {
+            let found = trusted_peers
+                .read()
+                .unwrap()
+                .iter()
+                .any(|(_peer_id, public_keys)| public_keys.identity_public_key == their_public_key);
+            if !found {
+                // TODO: security logging (mimoo)
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!(
+                        "noise: client connecting to us with an unknown public key: {}",
+                        their_public_key
+                    ),
+                ));
+            }
+        }
+
+        // if on a mutually authenticated network
+        if let Some(anti_replay_timestamps) = &anti_replay_timestamps {
+            // check that the payload received as the client timestamp (in seconds)
+            if payload.len() != PAYLOAD_SIZE {
+                // TODO: security logging (mimoo)
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "noise: client initiated connection without an 8-byte timestamp",
+                ));
+            }
+            let mut client_timestamp = [0u8; PAYLOAD_SIZE];
+            client_timestamp.copy_from_slice(&payload);
+            let client_timestamp = u64::from_le_bytes(client_timestamp);
+
+            // check the timestamp is not a replay
```

Comment:
1. How does the replay detection interact with a retry? In particular, if the client for some reason does not receive the responser's message, is there a chance that its retry of the Handshake would be flagged as a replay?
2. If the answer to the previous question is yes, how do you recommend we proceed on responder message loss?
3. Is this replay detection meant to guard against the server cost of performing a few DH operations, or is it meant to protect the server against client impersonation?

Reply:
1. good question! We thought about that so we made the timestamp accurate to the millisecond (and retries happen after X ms, with X configurable)

3. cost of performing a few DH operations

Reply:
As a client implementer, where do I find the minimal latency for retries?